apps/group/contracts/Group.sol

```diff
@@ -45,26 +65,78 @@ contract Group is App, Initializable, IForwarder, EVMCallScriptRunner {
     */
     function removeMember(address _entity) auth(REMOVE_MEMBER_ROLE) external {
         require(isGroupMember(_entity));
-        members[_entity] = false;
+        isMembers[_entity] = false;
+        for (uint i = 0; i < members.length; i++) {
+            if (members[i] == _entity) {
+                members[i] = members[members.length - 1];
+                delete members[members.length - 1];
+                members.length -= 1;
+                break;
+            }
+        }
+
+        if (required > members.length)
+            changeRequirement(members.length);
+
         RemoveMember(_entity);
     }
 
+    function numMembers() public constant returns (uint) {
+        return members.length;
+    }
+
+    function getRequired() public constant returns (uint) {
+        return required;
+    }
+
+    function confirm() public {
+        require(isGroupMember(msg.sender));
+        confirmations[msg.sender] = true;
+        Confirmation(msg.sender);
+    }
+
+    function revoke() public {
+        require(isGroupMember(msg.sender));
+        confirmations[msg.sender] = false;
+        Revocation(msg.sender);
+    }
+
+    function isConfirmed() public constant returns (bool) {
+        uint count = 0;
+        for (uint i = 0; i < members.length; i++) {
+            if (confirmations[members[i]]) {
+                count += 1;
+            }
+        }
+
+        return count >= required;
+    }
+
     /**
     * @dev IForwarder interface conformance. Forwards actions to any group member.
     * @param _evmCallScript Script being forwarded
     */
     function forward(bytes _evmCallScript) external {
-        require(isGroupMember(msg.sender));
+        require(isGroupMember(msg.sender) && isConfirmed());
```

Comment:
After adopting a confirmations per action model, forwarding should confirm the action and then if the action has enough confirmations, perform the action.